sklearn/benchmark.py

```diff
@@ -0,0 +1,171 @@
+    time_budget : int, default=300
+        Maximum number of seconds to use overall.  Current fit will
+        be stopped if the budget is exceeded.
+    profile_memory : bool, default=True
+        Whether to include memory (or just time) profiling. Memory
+        profiling will slow down fitting, and hence make fit_time
+        estimates more approximate.
+
+    Returns
+    -------
+    results : dict
+        The following keys are each mapped to an array:
+
+        n_samples
+            The number of samples
+        fit_time
+            In seconds
+        peak_memory
+            The memory used at peak of fitting, in KiB.
+        model_memory
+            The memory in use at the end of fitting, minus that at the
+            beginning, in KiB.
+
+    models : dict
+        keys 'peak_memory', 'model_memory' and 'fit_time' map to polynomial
+        GP regressors whose input is n_samples  and whose
+        outputs are each of those targets.
+
+    errors : list of dicts
+        lists the parameters that resulted in exceptions
+
+    Notes
+    -----
+    Memory estimates are only for the current process, memory will not be
+    estimated correctly if `n_jobs` with multiprocessing is used in the
+    estimator.
+
+    Peak memory measurements are approximate, as they are based on polling
+    current memory usage during execution.
+
+    """
+    fit_times = []
+    length = _num_samples(X)
+    parameters = est_params
+    n_samples = []
+    model_memory = []
+    peak_memory = []
+    models = {}
+    results = {}
+    errors = {}
+    errors["n_samples"] = []
+    SAMPLES = 8
+
+    if profile_memory:
+        try:
+            from memory_profiler import memory_usage
+        except ImportError:
+            raise ImportError("To run the profiler with `profile_memory=True,`"
+                              " you need to install 'memory_profiler`. Please "
+                              "run `pip install memory_profiler`.")
+
+    if parameters is not None:
+        est.set_params(**parameters)
+    if fit_params is None:
+        fit_params = {}
+
+    for _ in range(0, n_fits):
+        # if vary_n_samples=True, number of samples starts at 8 and doubles
+        # for every fit.
+        if vary_n_samples:
+            if 8*2**_ < length:
+                SAMPLES = 8*2**_
+            else:
+                SAMPLES = length
+
+            n_samples.append(SAMPLES)
+            train_size = n_samples[_] - 1
+            X_fit = X[:train_size]
+            y_fit = y[:train_size]
+        # if vary_n_samples=False, the estimator is fit on the entire data for
+        # every fit.
+        else:
+            n_samples.append(length)
+            X_fit = X[:length]
+            y_fit = y[:length]
+
+        start_time = time.time()
+
+        try:
+            if profile_memory:
+                args = (est, X_fit, y_fit, fit_params)
+                mem = memory_usage((est_arg_handler, args), interval=.0001)
```

Comment:
This seems to be unreasonably small interval, given the overhead in polling. Yes, I recognise that without a small enough interval, we are likely to miss the peak. I suppose for a very short fit, we could try rerunning with a smaller interval, but I think that's not necessary and we're better off just saying that for short fits the peak memory estimate is poor. We can assume that anything being profiled will run for multiple seconds (or a meaningful portion of the budget) at least once the number of samples gets large enough.